qcp/matrix.py

```diff
@@ -0,0 +1,150 @@
+# Copyright 2022 Tiernan8r
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+from abc import ABC
+from typing import Union
+from ._types import SCALARS, VECTOR, MATRIX
+
+class Matrix(ABC):
+    """
+    Method stubs for an immutable implementation of a matrix.
+    """
+
+    def __init__(self, state: MATRIX):
+        self.state = state
+        self.conjugated = False
+    
+    def dim(self):
+    #Return the dimension of matrix, in (row,col) tuple
+    #__len__ cannot return tuple, only integer
+        if not isinstance(self.state[0], list):
```

Comment:
Ideally, should have a check on the dimensions of `self.state` before indexing it, in case it is empty.